scale-decode-derive/src/lib.rs

```diff
@@ -260,48 +275,51 @@ fn generate_struct_impl(
                 type Error = #path_to_scale_decode::Error;
                 type Value<'scale, 'info> = #path_to_type #ty_generics;
+                type TypeResolver = #type_resolver_ident;
 
                 fn visit_composite<'scale, 'info>(
                     self,
-                    value: &mut #path_to_scale_decode::visitor::types::Composite<'scale, 'info>,
-                    type_id: #path_to_scale_decode::visitor::TypeId,
+                    value: &mut #path_to_scale_decode::visitor::types::Composite<'scale, 'info, Self::TypeResolver>,
+                    type_id: &<Self::TypeResolver as #path_to_scale_decode::TypeResolver>::TypeId,
                 ) -> Result<Self::Value<'scale, 'info>, Self::Error> {
                     #visit_composite_body
                 }
                 fn visit_tuple<'scale, 'info>(
                     self,
-                    value: &mut #path_to_scale_decode::visitor::types::Tuple<'scale, 'info>,
-                    type_id: #path_to_scale_decode::visitor::TypeId,
+                    value: &mut #path_to_scale_decode::visitor::types::Tuple<'scale, 'info, Self::TypeResolver>,
+                    type_id: &<Self::TypeResolver as #path_to_scale_decode::TypeResolver>::TypeId,
                 ) -> Result<Self::Value<'scale, 'info>, Self::Error> {
                     #visit_tuple_body
                 }
             }
 
-            impl #impl_generics #path_to_scale_decode::DecodeAsFields for #path_to_type #ty_generics #where_clause  {
-                fn decode_as_fields<'info>(input: &mut &[u8], fields: &mut dyn #path_to_scale_decode::FieldIter<'info>, types: &'info #path_to_scale_decode::PortableRegistry)
-                    -> Result<Self, #path_to_scale_decode::Error>
+            impl #impl_generics #path_to_scale_decode::DecodeAsFields for #path_to_type #ty_generics #visitor_where_clause  {
+                fn decode_as_fields<'info, R: #path_to_scale_decode::TypeResolver>(
+                    input: &mut &[u8],
+                    fields: &mut dyn #path_to_scale_decode::FieldIter<'info, R::TypeId>,
+                    types: &'info R
+                ) -> Result<Self, #path_to_scale_decode::Error>
                 {
-                    let path = #path_to_scale_decode::EMPTY_SCALE_INFO_PATH;
-                    let mut composite = #path_to_scale_decode::visitor::types::Composite::new(input, path, fields, types, false);
+                    let mut composite = #path_to_scale_decode::visitor::types::Composite::new(input, fields, types, false);
                     use #path_to_scale_decode::{ Visitor, IntoVisitor };
-                    let val = <#path_to_type #ty_generics>::into_visitor().visit_composite(&mut composite, #path_to_scale_decode::visitor::TypeId(0));
+                    let val = <#path_to_type #ty_generics>::into_visitor().visit_composite(&mut composite, &Default::default());
```

Comment:
Using a type ID of 0 was a bit of a hack before, so here and in one other place we change it to being `Default::default()` instead. This requires our `TypeId` to have a `Default` bound on it, which ideally it would do without.

In the future I expect that we can tweak these cases and remove the bound entirely, but for now I don't suspect it will be a big deal.